[Python] Incorrect argument type not flagged

Consider the following program where `foo` is called with an incorrect argument type:
```python
def foo(s: str) -> None:
    pass

foo(42)
```
Despite the type checking warning, verification succeeds:
```bash
$ esbmc ex.py 
ESBMC version 7.11.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
Parsing ex.py

Type checking warning:
ex.py:4: error: Argument 1 to "foo" has incompatible type "int"; expected "str"  [arg-type]
Found 1 error in 1 file (checked 1 source file)

Converting
Generating GOTO Program
GOTO program creation time: 0.549s
GOTO program processing time: 0.002s
Starting Bounded Model Checking
Symex completed in: 0.001s (2 assignments)
Caching time: 0.000s (removed 0 assertions)
Slicing time: 0.000s (removed 2 assignments)
Generated 0 VCC(s), 0 remaining after simplification (0 assignments)
BMC program time: 0.001s

VERIFICATION SUCCESSFUL
```
Is there a way this error can be captured/flagged?

--- a/regression/python/github_3020/main.py
+++ b/regression/python/github_3020/main.py
@@ -1,0 +1,4 @@
+def foo(s: str) -> None:
+    pass
+
+foo(42)

--- a/regression/python/github_3020/test.desc
+++ b/regression/python/github_3020/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--strict-types
+^VERIFICATION FAILED$

--- a/regression/python/github_3020_10/main.py
+++ b/regression/python/github_3020_10/main.py
@@ -1,0 +1,6 @@
+class C:
+    @classmethod
+    def make(cls, value: int) -> None:
+        pass
+
+C.make("wrong")

--- a/regression/python/github_3020_10/test.desc
+++ b/regression/python/github_3020_10/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--strict-types
+^VERIFICATION FAILED$

--- a/regression/python/github_3020_2/main.py
+++ b/regression/python/github_3020_2/main.py
@@ -1,0 +1,5 @@
+def bar(x: int, y: str, z: float) -> None:
+    pass
+
+bar("wrong", 123, "also wrong")
+

--- a/regression/python/github_3020_2/test.desc
+++ b/regression/python/github_3020_2/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--strict-types
+^VERIFICATION FAILED$

--- a/regression/python/github_3020_3/main.py
+++ b/regression/python/github_3020_3/main.py
@@ -1,0 +1,5 @@
+def add(a: int, b: int) -> int:
+    return a + b
+
+result = add(5, 10)
+assert result == 15

--- a/regression/python/github_3020_3/test.desc
+++ b/regression/python/github_3020_3/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--strict-types
+^VERIFICATION SUCCESSFUL$

--- a/regression/python/github_3020_4/main.py
+++ b/regression/python/github_3020_4/main.py
@@ -1,0 +1,4 @@
+def greet(name: str) -> None:
+    pass
+
+greet(42)

--- a/regression/python/github_3020_4/test.desc
+++ b/regression/python/github_3020_4/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+
+^VERIFICATION SUCCESSFUL$

--- a/regression/python/github_3020_5/main.py
+++ b/regression/python/github_3020_5/main.py
@@ -1,0 +1,4 @@
+def process(value: int) -> None:
+    pass
+
+process(3.14)

--- a/regression/python/github_3020_5/test.desc
+++ b/regression/python/github_3020_5/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--strict-types
+^VERIFICATION FAILED$

--- a/regression/python/github_3020_6/main.py
+++ b/regression/python/github_3020_6/main.py
@@ -1,0 +1,6 @@
+class Calculator:
+    def multiply(self, x: int, y: int) -> int:
+        return x * y
+
+calc = Calculator()
+calc.multiply(5, "ten")

--- a/regression/python/github_3020_6/test.desc
+++ b/regression/python/github_3020_6/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--strict-types
+^VERIFICATION FAILED$

--- a/regression/python/github_3020_7/main.py
+++ b/regression/python/github_3020_7/main.py
@@ -1,0 +1,7 @@
+class Calculator:
+    def add(self, x: int, y: int) -> int:
+        return x + y
+
+calc = Calculator()
+result = calc.add(3, 7)
+assert result == 10

--- a/regression/python/github_3020_7/test.desc
+++ b/regression/python/github_3020_7/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--strict-types
+^VERIFICATION SUCCESSFUL$

--- a/regression/python/github_3020_8/main.py
+++ b/regression/python/github_3020_8/main.py
@@ -1,0 +1,5 @@
+def process(value: int) -> int:
+    return value * 2
+
+result1 = process(5)  # Correct
+result2 = process("wrong")  # Should fail here

--- a/regression/python/github_3020_8/test.desc
+++ b/regression/python/github_3020_8/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--strict-types
+^VERIFICATION FAILED$

--- a/regression/python/github_3020_9/main.py
+++ b/regression/python/github_3020_9/main.py
@@ -1,0 +1,6 @@
+class MyClass:
+    @staticmethod
+    def static_func(x: int) -> int:
+        return x * 2
+
+MyClass.static_func("not an int")

--- a/regression/python/github_3020_9/test.desc
+++ b/regression/python/github_3020_9/test.desc
@@ -1,4 +1,4 @@
-KNOWNBUG
+CORE
 main.py
 --strict-types
 ^VERIFICATION FAILED$

--- a/regression/python/github_3020_9/test.desc
+++ b/regression/python/github_3020_9/test.desc
@@ -1,0 +1,4 @@
+KNOWNBUG
+main.py
+--strict-types
+^VERIFICATION FAILED$

--- a/src/esbmc/options.cpp
+++ b/src/esbmc/options.cpp
@@ -75,6 +75,10 @@ const struct group_opt_templ all_cmd_options[] = {
      {"override-return-annotation",
       NULL,
       "Override return annotation with inferred type"},
+     {"strict-types",
+      NULL,
+      "Enforce strict type checking for function arguments during "
+      "verification"},
    }},
 #endif
 #ifdef ENABLE_SOLIDITY_FRONTEND

--- a/src/python-frontend/function_call_expr.cpp
+++ b/src/python-frontend/function_call_expr.cpp
@@ -1573,14 +1573,6 @@ exprt function_call_expr::handle_general_function_call()
   // Find function symbol
   const symbolt *func_symbol = converter_.find_symbol(func_symbol_id);
 
-  if (func_symbol != nullptr)
-  {
-    // Add type checking here
-    exprt type_error = check_argument_types(func_symbol, call_["args"]);
-    if (!type_error.is_nil())
-      return type_error;
-  }
-
   if (func_symbol == nullptr)
   {
     if (
@@ -1798,6 +1790,13 @@ exprt function_call_expr::handle_general_function_call()
         }
       }
     }
+  }
+
+  if (func_symbol != nullptr)
+  {
+    exprt type_error = check_argument_types(func_symbol, call_["args"]);
+    if (!type_error.is_nil())
+      return type_error;
   }
 
   locationt location = converter_.get_location_from_decl(call_);
@@ -2214,12 +2213,26 @@ exprt function_call_expr::check_argument_types(
   const code_typet &func_type = to_code_type(func_symbol->type);
   const auto &params = func_type.arguments();
 
-  // Skip 'self'/'cls' parameter for instance/class methods
+  // Determine parameter offset based on actual function signature
   size_t param_offset = 0;
-  if (
-    function_type_ == FunctionType::InstanceMethod ||
-    function_type_ == FunctionType::ClassMethod)
+
+  if (function_type_ == FunctionType::InstanceMethod)
+  {
+    // Instance methods always have 'self' as first parameter
     param_offset = 1;
+  }
+  else if (function_type_ == FunctionType::ClassMethod)
+  {
+    // For class methods, check if first parameter is actually 'self' or 'cls'
+    // Static methods are called as ClassMethod but don't have self/cls
+    if (!params.empty())
+    {
+      const std::string &first_param = params[0].get_base_name().as_string();
+      if (first_param == "self" || first_param == "cls")
+        param_offset = 1;
+      // Otherwise it's a static method, param_offset stays 0
+    }
+  }
 
   for (size_t i = 0; i < args.size(); ++i)
   {

--- a/src/python-frontend/function_call_expr.cpp
+++ b/src/python-frontend/function_call_expr.cpp
@@ -1573,6 +1573,14 @@ exprt function_call_expr::handle_general_function_call()
   // Find function symbol
   const symbolt *func_symbol = converter_.find_symbol(func_symbol_id);
 
+  if (func_symbol != nullptr)
+  {
+    // Add type checking here
+    exprt type_error = check_argument_types(func_symbol, call_["args"]);
+    if (!type_error.is_nil())
+      return type_error;
+  }
+
   if (func_symbol == nullptr)
   {
     if (
@@ -2193,4 +2201,56 @@ exprt function_call_expr::generate_attribute_error(
   assert_code.location().comment(error_msg.str());
 
   return assert_code;
+}
+
+exprt function_call_expr::check_argument_types(
+  const symbolt *func_symbol,
+  const nlohmann::json &args) const
+{
+  // Only perform type checking if --strict-types is enabled
+  if (!config.options.get_bool_option("strict-types"))
+    return nil_exprt();
+
+  const code_typet &func_type = to_code_type(func_symbol->type);
+  const auto &params = func_type.arguments();
+
+  // Skip 'self'/'cls' parameter for instance/class methods
+  size_t param_offset = 0;
+  if (
+    function_type_ == FunctionType::InstanceMethod ||
+    function_type_ == FunctionType::ClassMethod)
+    param_offset = 1;
+
+  for (size_t i = 0; i < args.size(); ++i)
+  {
+    size_t param_idx = i + param_offset;
+    if (param_idx >= params.size())
+      break;
+
+    exprt arg = converter_.get_expr(args[i]);
+    const typet &expected_type = params[param_idx].type();
+    const typet &actual_type = arg.type();
+
+    // Check for type mismatch
+    if (!base_type_eq(expected_type, actual_type, converter_.ns))
+    {
+      std::string expected_str = type_handler_.type_to_string(expected_type);
+      std::string actual_str = type_handler_.type_to_string(actual_type);
+
+      std::ostringstream msg;
+      msg << "TypeError: Argument " << (i + 1) << " has incompatible type '"
+          << actual_str << "'; expected '" << expected_str << "'";
+
+      exprt exception = gen_exception_raise("TypeError", msg.str());
+
+      // Add location information from the call
+      locationt loc = converter_.get_location_from_decl(call_);
+      exception.location() = loc;
+      exception.location().user_provided(true);
+
+      return exception;
+    }
+  }
+
+  return nil_exprt();
 }

--- a/src/python-frontend/function_call_expr.h
+++ b/src/python-frontend/function_call_expr.h
@@ -37,6 +37,14 @@ public:
   }
 
 private:
+  /*
+   * Validates that function call arguments match expected parameter types.
+   * Returns TypeError exception if type mismatch is detected, nil_exprt otherwise.
+   */
+  exprt check_argument_types(
+    const symbolt *func_symbol,
+    const nlohmann::json &args) const;
+
   // Helper methods for AttributeError detection
   std::vector<std::string>
   find_possible_class_types(const symbolt *obj_symbol) const;

